explorer-api/src/ping/models.rs

```diff
@@ -0,0 +1,61 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::{Duration, SystemTime};
+
+use serde::Deserialize;
+use serde::Serialize;
+use tokio::sync::RwLock;
+
+use mixnet_contract::IdentityKey;
+
+pub(crate) type PingCache = HashMap<IdentityKey, PingCacheItem>;
+
+#[derive(Clone)]
+pub(crate) struct ThreadsafePingCache {
+    inner: Arc<RwLock<PingCache>>,
+}
+
+impl ThreadsafePingCache {
+    pub(crate) fn new() -> Self {
+        ThreadsafePingCache {
+            inner: Arc::new(RwLock::new(PingCache::new())),
+        }
+    }
+
+    pub(crate) async fn get(&self, identity_key: IdentityKey) -> Option<PingResponse> {
+        self.inner
+            .read()
+            .await
+            .get(&identity_key)
+            .and_then(|cache_item| {
+                if cache_item.valid_until.gt(&SystemTime::now()) {
```

Comment:
I think we can also use our `filter` and `map` approach here

Reply:
Done 😀